error-prone-utils/src/main/java/tech/picnic/errorprone/utils/SourceCode.java

```diff
@@ -42,6 +43,27 @@ public static String treeToString(Tree tree, VisitorState state) {
     return src != null ? src : tree.toString();
   }
 
+  /**
+   * Returns a Java string constant expression (i.e., a quoted string) representing the given input.
+   *
+   * @apiNote This method differs from {@link com.sun.tools.javac.util.Constants#format(Object)} in
+   *     that it does not superfluously escape single quote characters.
+   * @param str The string of interest.
+   * @return A non-{@code null} string.
+   */
+  public static String toStringConstantExpression(CharSequence str) {
+    StringBuilder result = new StringBuilder("\"");
+    for (int i = 0; i < str.length(); i++) {
+      char c = str.charAt(i);
+      if (c == '\'') {
+        result.append('\'');
+      } else {
+        result.append(Convert.quote(c));
+      }
+    }
+    return result.append('"').toString();
+  }
```

Comment:
Just noticed we could also use the upstream [`VisitorState#getConstantExpression`](https://github.com/google/error-prone/blob/04c76af83debe4c507c842fcf22bf8d0ddfd9c58/check_api/src/main/java/com/google/errorprone/VisitorState.java#L777-L801) method, though this variant is more efficient.

Reply:
I filed google/error-prone#4586; will update this PR.

Reply:
Rebased and added a commit.